core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingThreadPoolWDLParser.scala

```diff
@@ -0,0 +1,119 @@
+  private val cacheTTLFailure = Duration(wdlParsingConf.getInt("cache-ttl-failure-seconds"), TimeUnit.SECONDS)
+  private val threadPoolSize = wdlParsingConf.getInt("parser-thread-pool-size")
+  private val threadPoolTimeout = wdlParsingConf.getInt("parser-thread-pool-timeout-seconds")
+
+  // set up cache for WDL parsing
+  /* from scalacache doc: "Note: If you’re using an in-memory cache (e.g. Guava or Caffeine) then it makes sense
+     to use the synchronous mode. But if you’re communicating with a cache over a network (e.g. Redis, Memcached)
+     then this mode is not recommended. If the network goes down, your app could hang forever!"
+   */
+  import scalacache.modes.sync._
+  private val underlyingCaffeineCache = Caffeine.newBuilder()
+    .maximumSize(cacheMaxSize)
+    .build[String, Entry[Try[ParsedWdlWorkflow]]]
+  implicit val customisedCaffeineCache: Cache[Try[ParsedWdlWorkflow]] = CaffeineCache(underlyingCaffeineCache)
+
+  // set up thread pool
+  val executorService: ExecutorService = Executors.newFixedThreadPool(threadPoolSize)
+
+  def parse(wdl: String): Try[ParsedWdlWorkflow] = {
+    val tick = System.currentTimeMillis()
+    val key = generateCacheKey(wdl)
+    logger.debug(s"<parseWDL-cache> looking up $key ...")
+    get(key) match {
+      case Some(parseResult) =>
+        val tock = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> found cached result for $key in $tock ms.")
+        parseResult
+      case None =>
+        val miss = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> encountered cache miss for $key in $miss ms.")
+        val parseResult = threadedParse(wdl)
+        val parsetime = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> actively parsed WDL for $key in $parsetime ms.")
+        val ttl = parseResult match {
+          case Success(_) => Some(cacheTTLSuccess)
+          case Failure(ex) => ex.getCause match {
+            case se:SyntaxError =>
+              // syntax error is an expected, deterministic response to invalid wdl. cache this equivalent to a success.
+              Some(cacheTTLSuccess)
+            case _  =>
+              // other errors may be transient, such as timeouts retrieving http imports, or timeouts
+              // on the thread pool because parsing was slow
+              Some(cacheTTLFailure)
+          }
+        }
+        put(key)(parseResult, ttl = ttl)
+        val tock = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> cached result $key in $tock ms.")
+        parseResult
+    }
+  }
+
+  // an alternate terse/idiomatic implementation, but can't get good logging
+  /*
+  def parse(wdl: String): Try[ParsedWdlWorkflow] = {
+    sync.caching(generateCacheKey(wdl))(ttl = None) {
+      threadedParse(wdl)
+    }
+  }
+  */
+
+  private def threadedParse(wdl: String): Try[ParsedWdlWorkflow] = {
+    try {
```

Comment:
capital T Try then you don't need the catch

Reply:
for whatever reason I couldn't get this to compile; I think it's something about the nested `Try`s. I'll give it another shot.